### DISCUSSION TITLE
Proper way to reference extensibility package

### DISCUSSION THREAD

**Discussion by TaffarelJr** [xunit/xunit#2576] [Question] [answered]
I have a test utility project that I intend to share across several xUnit test projects.
One of the components I'm sharing is a test fixture.
In that fixture, I need an `IMessageSink` so I can output diagnostic info.
In order to write that info, I need to create a `Xunit.Sdk.DiagnosticMessage` object.
To get that object, I'm referencing the `xunit.extensibility.execution` NuGet package.

My test utility project compiles just fine. I referenced it in my xUnit test project (same solution), and when I run tests I can see it's getting into the fixture constructor. But as soon as the fixture constructor tries to create a `DiagnosticMessage`, it throws:

`System.IO.FileNotFoundException: 'Could not load file or assembly 'xunit.execution.dotnet, Version=2.4.2.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c' or one of its dependencies. The system cannot find the file specified.'`

I've been reading through the info on the various xUnit NuGet packages [here](https://xunit.net/docs/nuget-packages), but I can't figure out what I'm doing wrong. According to the documentation, the `xunit.extensibility.execution` package _should_:

> "[add] a reference to the platform specific execution DLL, rather than simply copying to the project's output folder,"

 ... shouldn't it? Why is my test utility project not including a transitive reference to the `dotnet` assembly?

**Comment by TaffarelJr** [comment, marked as answer]
I should add that my test utility library is multi-targeting:

- `netstandard2.0`
- `netstandard2.1`
- `net6.0`

... and the xUnit test project that references it is multi-targeting:

- `net462` — _Fails to find the `xunit.execution.dotnet` library_
- `net47` — _Fails to find the `xunit.execution.dotnet` library_
- `net471` — _Fails to find the `xunit.execution.dotnet` library_
- `net472` — _Fails to find the `xunit.execution.dotnet` library_
- `net48` — _Fails to find the `xunit.execution.dotnet` library_
- `netcoreapp3.1` — _Succeeds_
- `net5.0` — _Succeeds_
- `net6.0` — _Succeeds_

So it's only the tests running under the .NET Framework where the fixture fails to find the `xunit.execution.dotnet` library.

**Reply by bradwilson** [reply]
This is because `xunit.extensibility.execution` supports exactly `netstandard1.1` and `net452`, and those dependent DLLs have different names: `xunit.execution.dotnet.dll` and `xunit.execution.desktop.dll`.

The correct DLL is chosen at runtime based on whether your tests target .NET Framework (`xunit.execution.desktop`) or .NET Core (`xunit.execution.dotnet`). Unfortunately, your multi-targeted test utility library does not have a version which explicitly targets .NET Framework, so your `net4x` tests are likely all pulling the `netstandard2.0` version of your test utility library, which itself references `xunit.execution.dotnet` (because you're consuming the `netstandard1.1` version of the package, not the `net452` version). As such, this is not compatible with .NET Framework test projects.

To fix this, you should only need to add a `net462` target for your test utility library so that your .NET Framework targeted tests reference that version instead of the `netstandard2.0` version (and thus get the correct `xunit.execution.desktop` reference that is required for .NET Framework projects).
